client/mysqldump.c

```diff
@@ -4165,6 +4252,397 @@ static char *getTableName(int reset)
+      opt_ignore ? "IF NOT EXISTS " : "", row[0]);
+    fprintf(md_result_file, "/*M!100005 GRANT %s TO %s%s*/;\n",
+            row[0], row[1], (row[2][0] == 'Y') ? " WITH ADMIN OPTION " : "");
+  }
+  mysql_free_result(tableres);
+
+  /* users and their default role */
+  if (maria_roles_exist && mysql_query_with_error_report(mysql, &tableres,
+      "select IF(default_role='', 'NONE', QUOTE(default_role)) as r,"
+      "concat(QUOTE(User), '@', QUOTE(Host)) as u FROM mysql.user  "
+      "/*M!100005 WHERE is_role='N' */"))
+    return 1;
+  if (mysql_roles_exist && mysql_roles_exist && mysql_query_with_error_report(mysql, &tableres,
+      "SELECT IF(DEFAULT_ROLE_HOST IS NULL, 'NONE', CONCAT(QUOTE(DEFAULT_ROLE_USER), '@', QUOTE(DEFAULT_ROLE_HOST))) as r,"
+      "  CONCAT(QUOTE(mu.USER),'@',QUOTE(mu.HOST)) as u "
+      "FROM mysql.user mu LEFT JOIN mysql.default_roles using (USER, HOST)"))
+    return 1;
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    if (dump_grants(row[1]))
+      result= 1;
+    fprintf(md_result_file, "/*M!100005 SET DEFAULT ROLE %s FOR %s */;\n", row[0], row[1]);
+    fprintf(md_result_file, "/*!80001 ALTER USER %s DEFAULT ROLE %s */;\n", row[1], row[0]);
+  }
+  mysql_free_result(tableres);
+
+  if (maria_roles_exist && mysql_query_with_error_report(mysql, &tableres,
+      "SELECT DISTINCT QUOTE(m.role) AS r "
+      "   FROM mysql.roles_mapping m"
+      "   JOIN mysql.user u ON u.user = m.role"
+      "   WHERE is_role='Y'"
+      "     AND Admin_option='Y'"
+      "   ORDER BY m.role"))
+    return 1;
+  if (mysql_roles_exist && mysql_query_with_error_report(mysql, &tableres,
+      "SELECT DISTINCT CONCAT(QUOTE(FROM_USER),'@', QUOTE(FROM_HOST)) AS r "
+      "FROM mysql.role_edges"))
+    return 1;
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    if (dump_grants(row[0]))
+      result= 1;
+  }
+  mysql_free_result(tableres);
+  /* switch back */
+  fputs("SET ROLE NONE;\n"
+        "DROP ROLE mariadb_dump_import_role;\n"
+        "/*M!100203 EXECUTE IMMEDIATE CONCAT('SET ROLE ', @current_role) */;\n",
+        md_result_file);
+exit:
+
+  return result;
+}
+
+
+/*
+  dump all plugins
+*/
+
+static int dump_all_plugins()
+{
+  MYSQL_ROW row;
+  MYSQL_RES *tableres;
+
+  if (mysql_query_with_error_report(mysql, &tableres, "SHOW PLUGINS"))
+    return 1;
+  /* Name, Status, Type, Library,License */
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    if (strcmp("ACTIVE", row[1]) != 0)
+      continue;
+    /* Should we be skipping builtins? */
+    if (row[3] == NULL)
+      continue;
+    fprintf(md_result_file,
+       "INSTALL PLUGIN %s %s SONAME '%s';\n", row[0], opt_ignore ? "/*!100401 IF NOT EXISTS */" : "", row[3]);
+  }
+  mysql_free_result(tableres);
+
+  return 0;
+}
+
+
+/*
+  dump all udf
+*/
+
+static int dump_all_udf()
+{
+  /* we don't support all these types yet, but get prepared if we do */
+  static const char *udf_types[] = {"STRING", "REAL", "INT", "ROW", "DECIMAL", "TIME" };
```

Comment:
ROW, TIME can be dropped, no need for this. If we support it we can add, now is making at least for me confusion as in server code for `Item_result`

Reply:
They are fixed because of the plugin api already. I don't trust our future selves to remember this. Also the number indexes into this array so a value string needs to be be there, even if its unlikely to be hit at the moment.